### DISCUSSION TITLE
Custom display / visualization?

### DISCUSSION THREAD

**Discussion by alexalemi** [observablehq/framework#1851] [Q&A] [answered]
I feel like I might be missing something, but is there a way to generate custom visualizations for my own javascript classes?

I'm looking for something like the [custom viewers in clerk notebooks](https://github.clerk.garden/nextjournal/book-of-clerk/commit/a343752adb110e80a7cb33e375517eb3597e9db1/#writing-viewers) or the custom rich display hooks like [`_repr_html_`](https://ipython.readthedocs.io/en/stable/config/integrating.html#rich-display) for jupyter notebooks.

I have my own javascript classes, and I'd like to create custom visualizations for them inside Observable Framework, how do I do that?

**Comment by Fil** [comment]
If you want to visualize anything in framework, you can `display()` it — and it will call the default inspector, which knows how to display arrays, strings, (nested) objects, images, DOM elements etc. If you need a custom visualization, you can create a function that returns something more useful (maybe an interactive chart).

For example, if you want to visualize a graph with nodes and links, you could use a function that returns an arc diagram, then call `display(arcDiagram(graph))` — https://observablehq.observablehq.cloud/pangea/plot/arc-diagram 

I'm not sure this answers your question, though?

**Reply by mbostock** [reply, marked as answer]
To add, you can combine the built-in `display` with libraries that generate DOM content, such as [D3](https://observablehq.com/framework/lib/d3), [Observable Plot](https://observablehq.com/framework/lib/plot), [Observable Inputs](https://observablehq.com/framework/inputs/), [Hypertext Literal](https://observablehq.com/framework/lib/htl), and [React](https://observablehq.com/framework/jsx). You can also generate your own DOM content using the standard [DOM API](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API).
